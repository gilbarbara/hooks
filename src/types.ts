import { RefObject } from 'react';

export type PlainObject<T = unknown> = Record<PropertyKey, T>;

export type Primitive = bigint | boolean | null | number | string | symbol | undefined;

export type Target<T extends Element> = RefObject<T> | T | null | string;

export type TimerStatus = 'pending' | 'completed' | 'cancelled';
