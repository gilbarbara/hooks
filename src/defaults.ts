import type { UseElementMeasureResult } from './useElementMeasure';

export const defaultElementDimensions: UseElementMeasureResult = {
  absoluteHeight: 0,
  absoluteWidth: 0,
  bottom: 0,
  height: 0,
  left: 0,
  right: 0,
  top: 0,
  width: 0,
  x: 0,
  y: 0,
};
