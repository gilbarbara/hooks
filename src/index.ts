export { useClickOutside } from './useClickOutside';
export { useDeepCompareEffect } from './useDeepCompareEffect';
export { useDebounce } from './useDebounce';
export { useEffectOnce } from './useEffectOnce';
export {
  useFetch,
  USE_FETCH_STATUS,
  type UseFetchOptions,
  type UseFetchResult,
  type UseFetchStatus,
} from './useFetch';
export { useIntersectionObserver } from './useIntersectionObserver';
export { useInterval } from './useInterval';
export { useIsFirstMount } from './useIsFirstMount';
export { useIsMounted } from './useIsMounted';
export { useIsomorphicLayoutEffect } from './useIsomorphicLayoutEffect';
export { useLatest } from './useLatest';
export { useLifecycles } from './useLifecycles';
export {
  useLocalStorage,
  type UseLocalStorageOptions,
  type UseLocalStorageResult,
} from './useLocalStorage';

export {
  useLocalStorageState,
  type UseLocalStorageStateOptions,
  type UseLocalStorageStateResult,
} from './useLocalStorageState';
export { useLocation, type UseLocationResult } from './useLocation';
export { useMeasure, type UseMeasureResult } from './useMeasure';
export { useMediaQuery } from './useMediaQuery';
export { useMergeRefs } from './useMergeRefs';
export { useMount } from './useMount';
export { usePrevious } from './usePrevious';
export { useRenderCount } from './useRenderCount';
export { useResizeObserver } from './useResizeObserver';
export {
  useResponsive,
  type UseResponsiveOrientation,
  type UseResponsiveResult,
} from './useResponsive';
export { useScript, type UseScriptResult } from './useScript';
export { useSetState } from './useSetState';
export { useSingleton } from './useSingleton';
export { useTimeout, type UseTimeoutStatus, type UseTimeoutResult } from './useTimeout';
export { useThrottle } from './useThrottle';
export { useThrottleValue } from './useThrottleValue';
export { useToggle, type UseToggleResult } from './useToggle';
export { useUnmount } from './useUnmount';
export { useUpdate } from './useUpdate';
export { useUpdateEffect } from './useUpdateEffect';
export {
  useWhyDidYouUpdate,
  type UseWhyDidYouUpdateOptions,
  type UseWhyDidYouUpdateResult,
} from './useWhyDidYouUpdate';
export { useWindowSize, type UseWindowSizeResult } from './useWindowSize';
