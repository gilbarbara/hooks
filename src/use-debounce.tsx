import { useEffect, useState } from 'react';

export default function useDebounce(value: unknown, ms: number): unknown {
  // State and setters for debounced value
  const [debouncedValue, setDebouncedValue] = useState(value);

  useEffect(
    () => {
      // Update debounced value after delay
      const handler = setTimeout(() => {
        setDebouncedValue(value);
      }, ms);

      // Cancel the timeout if value changes (also on delay change or unmount)
      // This is how we prevent debounced value from updating if value is changed ...
      // .. within the delay period. Timeout gets cleared and restarted.
      return () => {
        clearTimeout(handler);
      };
    },
    [value, ms], // Only re-call effect if value or delay changes
  );

  return debouncedValue;
}
