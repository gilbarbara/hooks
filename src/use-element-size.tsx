import { useEffect, useLayoutEffect, useMemo, useState } from 'react';

export type UseElementRect = Omit<DOMRectReadOnly, 'toJSON'>;

const defaultState: UseElementRect = {
  x: 0,
  y: 0,
  width: 0,
  height: 0,
  top: 0,
  left: 0,
  bottom: 0,
  right: 0,
};

export default function useElementSize(selector: string): UseElementRect {
  const [element, setElement] = useState<Element | null>();
  const [rect, setRect] = useState<UseElementRect>(defaultState);

  useEffect(() => {
    setElement(document.querySelector(selector));
  }, [selector]);

  const observer = useMemo(
    () =>
      new window.ResizeObserver(entries => {
        if (entries[0]) {
          const { bottom, height, left, right, top, width, x, y } = entries[0].contentRect;

          setRect({ x, y, width, height, top, left, bottom, right });
        }
      }),
    [],
  );

  useLayoutEffect(() => {
    if (!element) {
      return;
    }

    observer.observe(element);

    // eslint-disable-next-line consistent-return
    return () => {
      observer.disconnect();
    };
  }, [observer, element]);

  return rect;
}
