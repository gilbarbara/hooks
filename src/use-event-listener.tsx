import { useEffect, useRef, MutableRefObject } from 'react';

type Handler = (e: Event) => void;

export default function useEventListener(
  eventName: string,
  handler: Handler,
  element: Window | Document | HTMLElement = window,
) {
  const savedHandler: MutableRefObject<null | Handler> = useRef(null);

  useEffect(() => {
    savedHandler.current = handler;
  }, [handler]);

  useEffect(() => {
    const { current } = savedHandler;
    const eventListener = (event: Event) => {
      /* istanbul ignore else */
      if (current) {
        current!(event);
      }
    };

    element.addEventListener(eventName, eventListener);

    return () => {
      element.removeEventListener(eventName, eventListener);
    };
  }, [element, eventName]);
}
