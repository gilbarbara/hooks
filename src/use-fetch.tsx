import { useEffect, useRef, useState } from 'react';

import { PlainObject } from './types';
import { isPlainObject, isString, isURL } from './utils';

export type FetchStatus = 'idle' | 'running' | 'success' | 'failure';

interface ResponseError extends Error {
  response?: unknown;
  status?: number;
}

interface useFetchOptions {
  body?: BodyInit;
  headers?: PlainObject;
  method?: string;
  mode?: 'cors' | 'navigate' | 'no-cors' | 'same-origin';
  type?: 'json' | 'urlencoded';
  url: string;
}

interface useFetchResponse<T> {
  data?: T;
  error?: ResponseError;
  status: FetchStatus;
}

async function request(options: useFetchOptions): Promise<any> {
  const {
    headers = {},
    method = 'GET',
    mode = 'cors',
    body = undefined,
    type = 'json',
    url = '',
  } = options;

  const contentTypes = {
    json: 'application/json',
    urlencoded: 'application/x-www-form-urlencoded',
  };

  const params: RequestInit = {
    body: undefined,
    cache: 'no-store' as const,
    headers: {
      Accept: 'application/json',
      'Content-Type': type ? contentTypes[type] : 'application/json',
      ...headers,
    },
    method,
    mode,
    credentials: undefined,
  };

  if (body && !isString(body) && type === 'json') {
    params.body = JSON.stringify(body);
  } else {
    params.body = body;
  }

  return fetch(url, params).then(async response => {
    const text = await response.text();
    let content: unknown;

    try {
      content = JSON.parse(text);
    } catch {
      content = text;
    }

    if (response.status > 299) {
      const error = new Error(response.statusText) as ResponseError;

      error.status = response.status;
      error.response = content;

      throw error;
    }

    return content;
  });
}

export default function useFetch<T = unknown>(
  urlOrOptions: string | useFetchOptions,
  wait = false,
): useFetchResponse<T> {
  const isActive = useRef(false);
  const [state, setState] = useState<useFetchResponse<T>>({
    data: undefined,
    error: undefined,
    status: 'idle',
  });

  if (!isPlainObject(urlOrOptions) && !isURL(urlOrOptions)) {
    throw new Error('Expected an options object or URL');
  }

  useEffect(() => {
    isActive.current = true;

    return () => {
      isActive.current = false;
    };
  }, []);

  useEffect(() => {
    if (state.status === 'idle' && !wait) {
      setState(s => ({
        ...s,
        status: 'running',
      }));

      request(
        isURL(urlOrOptions)
          ? {
              type: 'json',
              url: urlOrOptions,
            }
          : urlOrOptions,
      )
        .then(data => {
          /* istanbul ignore else */
          if (isActive.current) {
            setState(s => ({
              ...s,
              data,
              status: 'success',
            }));
          }
        })
        .catch(error => {
          /* istanbul ignore else */
          if (isActive.current) {
            setState(s => ({
              ...s,
              error,
              status: 'failure',
            }));
          }
        });
    }
  }, [state, urlOrOptions, wait]);

  return state;
}
