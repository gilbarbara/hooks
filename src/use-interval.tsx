import { useCallback, useEffect, useRef, MutableRefObject } from 'react';

type Callback = () => void;

export default function useInterval(callback: Callback, ms: number) {
  const savedCallback: MutableRefObject<null | Callback> = useRef(null);
  const interval: MutableRefObject<null | number> = useRef(null);

  // Remember the latest callback.
  useEffect(() => {
    savedCallback.current = callback;
  }, [callback]);

  // Set up the interval.
  useEffect(() => {
    interval.current = window.setInterval(() => {
      /* istanbul ignore else */
      if (savedCallback.current) {
        savedCallback.current();
      }
    }, ms);

    return () => {
      clearInterval(interval.current!);
    };
  }, [ms]);

  return useCallback(() => {
    clearInterval(interval.current!);
  }, []);
}
