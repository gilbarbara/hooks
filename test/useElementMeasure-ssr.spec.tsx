/**
 * @vitest-environment node
 */

import * as React from 'react';
import { renderToString } from 'react-dom/server';

import { useElementMeasure } from '../src/useElementMeasure';

function Component() {
  const dimensions = useElementMeasure('#root');

  return <pre>{JSON.stringify(dimensions, null, 2)}</pre>;
}

describe('useElementMeasure-ssr', () => {
  it('should work without a DOM', () => {
    const view = renderToString(<Component />);

    expect(view).toMatchSnapshot();
  });
});
