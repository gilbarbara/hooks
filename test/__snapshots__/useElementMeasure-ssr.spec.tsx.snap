// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`useElementMeasure-ssr > should work without a DOM 1`] = `
"<pre>{
  &quot;absoluteHeight&quot;: 0,
  &quot;absoluteWidth&quot;: 0,
  &quot;bottom&quot;: 0,
  &quot;height&quot;: 0,
  &quot;left&quot;: 0,
  &quot;right&quot;: 0,
  &quot;top&quot;: 0,
  &quot;width&quot;: 0,
  &quot;x&quot;: 0,
  &quot;y&quot;: 0
}</pre>"
`;
